### ISSUE TITLE
Recording the path taken through the taskflow?

### ISSUE THREAD

**Issue by Levi-Armstrong** [taskflow/taskflow#225]
I currently leverage the tool available for generating a graph showing the the full taskflow and interested to learn if there is capability for recording the path taken through the graph when executed and visualize it in the graph?

**Comment by tsung-wei-huang** [comment]
@Levi-Armstrong do you mean you would like to highlight the critical paths or? We have [tfprof](https://taskflow.github.io/tfprof/) to visualize the execution timeline but not the graph or path execution along with the graph. However, this is definitely a feature I would like to consider or need people to contribute to.

**Comment by Levi-Armstrong** [comment]
I did not know about the tfprof, is there a wiki on how to use it? Related to your question, I am interested in the execution path taken and would like to highlight it on the graph. Do you have any guidance on how this may be implemented and I will see if I can get something prototyped?

**Comment by tsung-wei-huang** [comment]
Hi @Levi-Armstrong it is very easy to use: https://github.com/taskflow/tfprof  

You just need to invoke your program using `python3 tfprof.py your_program your_arguments` and paste the `output.tfp` (JSON) content to the [web visualizer](https://taskflow.github.io/tfprof/). The code is defined in [observer.hpp](https://github.com/taskflow/taskflow/blob/master/taskflow/core/observer.hpp). I will definitely be delighted to see any contributions to this.

**Comment by Levi-Armstrong** [comment]
Thank you. I will see what I can put together and report back.

**Comment by tsung-wei-huang** [comment]
No problem. Let me know if you would like to contribute to this visualization part - we can have another collaborative project on this front.

**Comment by Levi-Armstrong** [comment]
I am currently looking into this and interested to know if there is a way from the observer to know if there is a way to identify which task it is from the original taskflow? There is a name property but this is not unique, so looking for an alternative option.

**Comment by Levi-Armstrong** [comment]
Looking at how the dump function works it looks to leverage the raw pointer of the node as the unique identifier for building the original graph.

**Comment by tsung-wei-huang** [comment]
@Levi-Armstrong yes, ultimately `tf::Node` is the data structure to dig out all graph information. `tf::Task` and `tf::TaskView` are just a wrapper on top to prevent users from accessing important data internal to Taskflow. 

You are correct, the dump function contains all information, and I think in order to get the execution path (or critical path) we need some other info from the executor. All task execution are guarded by [this](https://github.com/taskflow/taskflow/blob/a40159c747bf93c1355a48ba63856b47ba9b2e9f/taskflow/core/executor.hpp#L274) and [this](https://github.com/taskflow/taskflow/blob/a40159c747bf93c1355a48ba63856b47ba9b2e9f/taskflow/core/executor.hpp#L275)

**Comment by Levi-Armstrong** [comment]
Would it be OK to expose a method in TaskView to get the underlining const raw point to the node? Then the observer would just need to store these visited nodes along with and integer for how many times the node was visited. Then one could expand the dump method or create a separate one to except this where it would build the full graph but annotate the ones visited along with how many times it was visited.

**Comment by tsung-wei-huang** [comment]
@Levi-Armstrong It is doable and I think you can also create your observer and make it a friend to `tf::Node`  or just modify the `TFProfObserver` to directly integrate your feature. Then we can work together to see how to visualize it on [tfprof](https://taskflow.github.io/tfprof/#) or other means.

**Comment by Levi-Armstrong** [comment]
That works. It may be nice to add a way for either getting the node or a UUID like what is shown below. 

```
std::uintptr_t getUUID() const
{
  return reinterpret_cast<std::uintptr_t>(_node);
}
```

**Comment by tsung-wei-huang** [comment]
@Levi-Armstrong I was thinking about to have a uuid for a node but it is going to take too much memory  at `tf::Node`. In your code, please keep in mind node may be recycled by Taskflow's object pool. So, two different nodes at different runs may have the same pointer value.

**Comment by Levi-Armstrong** [comment]
> In your code, please keep in mind node may be recycled by Taskflow's object pool. So, two different nodes at different runs may have the same pointer value.

I just want to make sure I understand what you are  describing. In the example below are you saying that task `A` will have a different node pointer for every call to `executer.run(taskflow)`? 

```
  tf::Executor executor;
  tf::Taskflow taskflow;

  auto [A, B, C, D] = taskflow.emplace(
    [] () { std::cout << "TaskA\n"; }, 
    [] () { std::cout << "TaskB\n"; }, 
    [] () { std::cout << "TaskC\n"; },
    [] () { std::cout << "TaskD\n"; }
  );                                           
                                                  
  A.precede(B);  // A runs before B 
  A.precede(C);  // A runs before C 
  B.precede(D);  // B runs before D 
  C.precede(D);  // C runs before D 
                                                     
  executor.run(taskflow).wait();
```

**Comment by tsung-wei-huang** [comment]
No. In your case, A remains unique and alive with its taskflow. I was referring to the same address being reused from a deleted taskflow. For example:

```cpp
// scope 1
{
  tf::Taskflow taskflow;
  auto A = taskflow.emplace([](){});   // object pool creates a node for A
}                                                                    // object pool recycle the node of A

{
  tf::Taskflow taskflow;
  auto B = taskflow.emplace([](){});  // object pool reuses the address which was used for A in scope 1
}
```

**Comment by Levi-Armstrong** [comment]
That makes sense, so maybe not call it `getUUID`` but `getID` which will allow a user to sync data generated from the observer with the tasks for a given taskflow..

```
// The ID is only unique for the life of the task 
std::uintptr_t getID() const
{
  return reinterpret_cast<std::uintptr_t>(_node);
}
```

**Comment by tsung-wei-huang** [comment]
Yes. I agree.

**Comment by Levi-Armstrong** [comment]
Actually just noticed that a task has a `hash_value()` which does this. Should this work?

**Comment by tsung-wei-huang** [comment]
The method hashes the underlying pointer value of the node address.

**Comment by tsung-wei-huang** [comment]
As a side note, I have largely simplified the generation of profiling data in the master branch - which will be soon released as v3. You can check this [page](https://taskflow.github.io/taskflow/Profiler.html). It is as simple as:

```bash
~$ TF_ENABLE_PROFILER=my_program.json ./my_program
```

**Comment by Levi-Armstrong** [comment]
Great does this mean that you no longer have to add the observer?

**Comment by tsung-wei-huang** [comment]
Observer remains there. I create another lightweight layer on top of observer and make the profiling process much much much more easier - :)
